### DISCUSSION TITLE
proper docs

### DISCUSSION THREAD

**Discussion by felixroos** [tidalcycles/strudel#92] [Ideas]
At some point, it would be good to have a proper docs website that is more than just a giant document. Especially as the API grows, a single document is too large.
Similar to tidal docs, we could use docusaurus for that, as it also supports MDX, which allows embedding the MiniREPL.

Probably wait for https://github.com/tidalcycles/strudel/issues/91

**Comment by yaxu** [comment]
How about embedding some docs in the source, is that often done in javascript land?

**Reply by felixroos** [reply]
yes, it would be good to have a little doc comment over each function. I also thought we could parse those out and generate an API doc, which should be a seperate thing from the tutorial

**Reply by yaxu** [reply]
I think we should add docusaurus to the monorepo then. I think it's possible to have more than one github pages site, by having the build folder a submodule. That said, probably best to build to `/docs`,  and move the tutorial into docusaurus?

**Comment by yaxu** [comment]
It would be good to move on this so as to not get too behind with documenting all these functions. I'm not good at the documentation habit but can imagine trying to add a simple test and documentation file for each function as a minimum.

I'm not sure how to quickly get started though. Maybe just add a vanilla docusaurus site under /docusaurus and start adding stuff to that, one file per function/set of functions as reference material? Then decide how to integrate with tutorial etc later?

**Reply by felixroos** [reply]
yes, it's time to catch up with documentation. It might be smart to document each function in source using jsdoc and converting that to docusaurus later, as the API docs. In addition, we can add more tutorial / guide like pages to the docs

**Comment by felixroos** [comment]
just tested jsdoc in in-source-doc branch. It allows generating api docs easily. though its type system is inferior to typescript. It can either render to html or markdown, and it should integrate easily with docusaurus. here is some example output: https://github.com/tidalcycles/strudel/blob/in-source-doc/api.md , generated from these changes https://github.com/tidalcycles/strudel/commit/2348d056cbd91a0ef02f5c48a5608c21718b6b58#diff-c8529a9530742d567d93fc894a928043a5356ff50e608d4130f80f5e6d4ac3de

**Comment by felixroos** [comment]
just tried jsdoc-json, which allows turning source comments into an easy to process json file: https://github.com/tidalcycles/strudel/blob/e0c77a5fa8d91c990d58f71e4ecb1e901177d2f6/doc.json#L1 that could be used to show dynamic tooltips in the editor, e.g. show the description of the params of a function while typing (by combining the user code AST with the params field)
